deltacat/compute/compactor_v2/utils/exception_handler.py

```diff
@@ -0,0 +1,138 @@
+import botocore
+import ray
+from ray.exceptions import (
+    RayError,
+    RuntimeEnvSetupError,
+    WorkerCrashedError,
+    NodeDiedError,
+    OutOfMemoryError,
+)
+from pyarrow.lib import ArrowException, ArrowInvalid, ArrowCapacityError
+from deltacat.exceptions import (
+    DependencyPyarrowError,
+    DependencyRayRuntimeSetupError,
+    DependencyPyarrowInvalidError,
+    DependencyPyarrowCapacityError,
+    DependencyDaftError,
+    DependencyDaftTransientError,
+    GeneralAssertionError,
+    DependencyRayWorkerDiedError,
+    DependencyRayOutOfMemoryError,
+    DependencyRayError,
+)
+from typing import Callable
+from deltacat.utils.ray_utils.runtime import (
+    get_current_ray_task_id,
+)
+from daft.exceptions import DaftTransientError, DaftCoreException
+import tenacity
+
+# List of retryable errors that can be automatically retried by Ray remote tasks
+RAY_TASK_RETRYABLE_ERROR_CODES = [
+    botocore.exceptions.ConnectionError,
+    botocore.exceptions.HTTPClientError,
+    ConnectionError,
+    TimeoutError,
+    DaftTransientError,
+    tenacity.RetryError,
+]
+
+
+def handle_exception(func: Callable):
+    def handle_compaction_exception(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except BaseException as e:
+            handle_compaction_step_exception(e)
+
+    return handle_compaction_exception
+
+
+def handle_compaction_step_exception(e: BaseException):
+    if isinstance(e, RayError):
+        _handle_ray_error(e)
+    if isinstance(e, ArrowException):
+        _handle_dependency_pyarrow_error(e)
+    elif isinstance(e, AssertionError):
+        _handle_assertion_error(e)
+    elif isinstance(e, DaftCoreException):
+        _handle_daft_error(e)
+    else:
+        raise e
```

Comment:
Every error must be converted to `DeltaCATError`, otherwise wrap it into `UnclassifiedDeltaCATError`. We shouldn't often see `UnclassifiedDeltaCATError`.